OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs

```diff
@@ -132,14 +132,39 @@ void IBotTick.BotTick(IBot bot)
 
 		bool ShouldBuildMCV()
 		{
+			// Only build MCV if we have a factory to build it with.
+			var enoughFactories = AIUtils.CountBuildingByCommonName(Info.McvFactoryTypes, player) > 0;
+			if (!enoughFactories)
+				return false;
+
 			// Only build MCV if we don't already have one in the field.
-			var allowedToBuildMCV = AIUtils.CountActorByCommonName(Info.McvTypes, player) == 0;
-			if (!allowedToBuildMCV)
+			var mcvCount = AIUtils.CountActorByCommonName(Info.McvTypes, player);
+			if (mcvCount > 0)
+				return false;
+
+			// Only build MCV if the conyardCount + mcvCount are lower than the minimum conyard count.
+			var conyardCount = AIUtils.CountBuildingByCommonName(Info.ConstructionYardTypes, player);
+			if (conyardCount + mcvCount >= Info.MinimumConstructionYardCount)
+				return false;
+
+			// Only build MCV if there isn't already one being built.
+			var blds = player.World.ActorsWithTrait<ProductionQueue>()
+				.Where(a => a.Actor.Owner == player && Info.McvFactoryTypes.Contains(a.Actor.Info.Name) && a.Trait.Enabled)
+				.Select(a => a.Trait);
+
+			int queuedMcvCount = 0;
+			foreach (var factory in blds)
+			{
+				var q = factory.AllQueued().Where(z => Info.McvTypes.Contains(z.Item));
+
+				if (q != null)
+					queuedMcvCount += q.Count();
```

Comment:
Seeing that we immediately return `false` in case there is a single mcv, this can simplify to something like:
```
for (var factory in queues)
	if (factory.AllQueued().Any(z => Info.McvTypes.Contains(z.Item)))
		return false;